crates/fuel-core/src/database/balances.rs

```diff
@@ -48,37 +59,101 @@ impl StorageMutate<ContractsAssets> for Database {
+        // value is the 32 bytes
+        tree.update(&(*key.contract_id()).into(), value.to_be_bytes().as_slice())
+            .map_err(|err| StorageError::Other(err.into()))?;
+
+        // Generate new metadata for the updated tree
+        let root = tree.root();
+        let metadata = SparseMerkleMetadata { root };
+        self.storage::<ContractsAssetsMerkleMetadata>()
+            .insert(key.contract_id(), &metadata)?;
+
+        prev
     }
 
     fn remove(
         &mut self,
         key: &<ContractsAssets as Mappable>::Key,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::remove(self, key.as_ref(), Column::ContractsAssets).map_err(Into::into)
+        let prev = Database::remove(self, key.as_ref(), Column::ContractsAssets)
+            .map_err(Into::into);
+
+        // Get latest metadata entry
+        let prev_metadata = self
+            .iter_all::<Vec<u8>, SparseMerkleMetadata>(
+                Column::ContractsAssetsMerkleMetadata,
+                Some(IterDirection::Reverse),
+            )
+            .next()
+            .transpose()?
+            .map(|(_, metadata)| metadata)
+            .unwrap_or_default();
+
+        let root = prev_metadata.root;
+        if root == [0; 32] {
+            // The tree is empty
+            return prev
+        }
+
+        // Load the tree saved in metadata
+        let storage = self.borrow_mut();
+        let mut tree: MerkleTree<ContractsAssetsMerkleData, _> =
+            MerkleTree::load(storage, &root)
+                .map_err(|err| StorageError::Other(err.into()))?;
+
+        // Update the key-value dataset. The key is the contract id and the
+        // value is the 32 bytes
+        tree.delete(&(*key.contract_id()).into())
+            .map_err(|err| StorageError::Other(err.into()))?;
+
+        // Generate new metadata for the updated tree
+        let root = tree.root();
+        let metadata = SparseMerkleMetadata { root };
+        self.storage::<ContractsAssetsMerkleMetadata>()
+            .insert(key.contract_id(), &metadata)?;
+
+        prev
     }
 }
 
 impl MerkleRootStorage<ContractId, ContractsAssets> for Database {
     fn root(&self, parent: &ContractId) -> Result<MerkleRoot, Self::Error> {
-        let items: Vec<(Vec<u8>, Word)> = Database::iter_all_by_prefix(
-            self,
-            Column::ContractsAssets,
-            Some(parent),
-            Some(IterDirection::Forward),
-        )
-        .try_collect()?;
-
-        let root = items
-            .iter()
-            .filter_map(|(key, value)| {
-                (&key[..parent.len()] == parent.as_ref()).then_some((key, value))
-            })
-            .sorted_by_key(|t| t.0)
-            .map(|(_, value)| value.to_be_bytes());
-
-        Ok(crypto::ephemeral_merkle_root(root).into())
+        let metadata = self
+            .storage::<ContractsAssetsMerkleMetadata>()
+            .get(parent)?;
+        let root = metadata
```

Comment:
Why in some places we use `[0; 32]` and in some `in_memory::MerkleTree::new().root()`? Maybe the implementation of the `Default` trait for `SparseMerkleMetadata` should use `in_memory::MerkleTree::new().root()` for the `root` field?

Reply:
The `Default` trait for `SparseMerkleMetadata` actually does use an empty SMT to get the root: https://github.com/FuelLabs/fuel-core/pull/1008/files#diff-7919c97bb9270707dfee96e03589ae3bd0474612f23f12b938a91ba95d81cd01R64. 

In some places, I just used `[0; 32]`, as a shortcut. I could use another empty SMT for the root instead, if that would be more clear.